src/persist-client/src/iter.rs

```diff
@@ -610,21 +643,31 @@ where
                         return None;
                     }
                 }
-                Some((k1, v1, t1)) => match &mut self.state {
-                    None => {
-                        self.state = part.pop(&mut self.parts);
-                    }
-                    Some((k0, v0, t0, d0)) => {
-                        if (*k0, *v0, &*t0) == (*k1, *v1, t1) {
-                            let (_, _, _, d1) = part
-                                .pop(&mut self.parts)
-                                .expect("popping from a non-empty iterator");
-                            d0.plus_equals(&d1);
-                        } else {
-                            break;
+                Some((k1, v1, t1)) => {
+                    match &mut self.state {
+                        None => {
+                            self.state = part.pop(&mut self.parts);
+                        }
+                        Some((k0, v0, t0, d0)) => {
+                            match (*k0, *v0, &*t0).cmp(&(*k1, *v1, t1)) {
+                                Ordering::Less => {
+                                    break;
+                                }
+                                Ordering::Equal => {
+                                    let (_, _, _, d1) = part
+                                        .pop(&mut self.parts)
+                                        .expect("popping from a non-empty iterator");
+                                    d0.plus_equals(&d1);
+                                }
+                                Ordering::Greater => {
+                                    // Don't want to log the entire KV, but it's interesting to know
+                                    // whether it's KVs going backwards or 'just' timestamps.
+                                    panic!("data arrived at the consolidator out of order (kvs equal? {})", (*k0, *v0) == (*k1, *v1));
```

Comment:
do we want to put this behind a feature flag at first, at least until we're relying on it for correctness? it'd be unfortunate if the release process finds some weird edge case that CI didn't catch and we have to put out a patch deploy

Reply:
Every _use_ of this code is behind a feature flag, so I feel relatively comfortable here. Does that sound like a reasonable safeguard or should we add a flag for just the assert?